### DISCUSSION TITLE
import both horovod.tensorflow and horovod.torch cause issues

### DISCUSSION THREAD

**Discussion by mengxr** [horovod/horovod#2752] [Q&A]
I hit an issue when the code import both horovod.tensorflow and horovod.torch and use the latter. It might not be a valid use case in batch jobs, but in interactive notebooks, it would cause issues. Is there a shared variable that causes this issue?

Env:
* Python 3.6
* TensorFlow 1.12
* PyTorch 0.4.1
* Horovod 0.15.0

Code:
test_torch.py
~~~python
import horovod.tensorflow
import torch
import horovod.torch as hvd

hvd.init()
hvd.broadcast(torch.tensor(0), root_rank=0)
~~~

Run:
~~~
$ python test_torch.py
~~~

Error:
~~~
Traceback (most recent call last):
  File "test_pytorch.sh", line 6, in <module>
    hvd.broadcast(torch.tensor(0), root_rank=0)
  File "/opt/conda/envs/sparkdl/lib/python3.6/site-packages/horovod/torch/mpi_ops.py", line 358, in broadcast
    return HorovodBroadcast.apply(tensor, root_rank, name)
  File "/opt/conda/envs/sparkdl/lib/python3.6/site-packages/horovod/torch/mpi_ops.py", line 324, in forward
    handle = broadcast_async(tensor, root_rank, name)
  File "/opt/conda/envs/sparkdl/lib/python3.6/site-packages/horovod/torch/mpi_ops.py", line 315, in broadcast_async
    return _broadcast_async(tensor, output, root_rank, name)
  File "/opt/conda/envs/sparkdl/lib/python3.6/site-packages/horovod/torch/mpi_ops.py", line 290, in _broadcast_async
    tensor, output, root_rank, name.encode() if name is not None else _NULL)
  File "/opt/conda/envs/sparkdl/lib/python3.6/site-packages/torch/utils/ffi/__init__.py", line 202, in safe_call
    result = torch._C._safe_call(*args, **kwargs)
torch.FatalError: Horovod has not been initialized; use hvd.init().
~~~

Without "import tensorflow.horovod", it runs successfully.

**Comment by tgaddair** [comment]
Hey @mengxr, thanks for raising this issue, though I wasn't able to repro the issue with the code provided.  Was there more to the script that triggered this error?  How did you execute the script (python, mpirun, notebook, etc.)?

**Comment by mengxr** [comment]
No `mpirun` was used. Just `python test_torch.py`

**Comment by tgaddair** [comment]
Thanks for the info, I'll try running this in a few more environments and see if I can reproduce it.

**Comment by mengxr** [comment]
I run this under docker "ubuntu:16.04" with miniconda. TensorFlow was installed from anaconda channel and pytorch-cpu was installed from pytorch channel (I also tried pip install whl directly). Horovod was installed from pip.

**Comment by alsrgv** [comment]
@mengxr, I wasn't able to reproduce this issue as well in current formulation.  Since you've mentioned docker, do you have a Dockerfile that you could share to speed up the repro?
